apollo-federation/src/sources/connect/json_selection/methods/public.rs

```diff
@@ -157,8 +230,68 @@ fn match_method(
         ),
     )
 }
+pub(super) fn match_shape(
+    method_name: &WithRange<String>,
+    method_args: Option<&MethodArgs>,
+    input_shape: Shape,
+    dollar_shape: Shape,
+    named_var_shapes: &IndexMap<&str, Shape>,
+) -> Shape {
+    if let Some(MethodArgs { args, .. }) = method_args {
+        let mut result_union = Vec::new();
+        let mut has_infallible_case = false;
+
+        for pair in args {
+            if let LitExpr::Array(pair) = pair.as_ref() {
+                if pair.len() == 2 {
```

Comment:
should we return an error if the length isn't exactly 2?

Reply:
do you have an example of error handling? since this doesn't return a `Result`, what should a caller do with a `Shape::error_with_range`?

Reply:
I imagine we'll add a way to collect/traverse/visit any `ShapeCase::Error` variants within a given `Shape`, to handle them programmatically or surface them in the debugger, but for now the main point of errors is to fail validation against expected GraphQL schema types.

Since you can embed `ShapeCase::Error` within a larger shape, shape errors are similar to the GraphQL errors-as-data pattern, where you're able to get back some partial shape information along with contextual errors.